runtime/src/main/java/org/corfudb/runtime/Restore.java

```diff
@@ -227,8 +234,71 @@ private void clearAllTables() {
         txn.commit();
     }
 
+    private void nonCachedAppendSMREntries(UUID streamId, SMREntry... smrEntries) {
+        MultiSMREntry multiSMREntry = new MultiSMREntry();
+        multiSMREntry.addTo(Arrays.asList(smrEntries));
+        rt.getStreamsView().append(multiSMREntry, null, CacheOption.WRITE_AROUND, streamId);
+    }
+
     public enum RestoreMode {
         FULL,   // Clean up ALL tables before restore
         PARTIAL // Clean up ONLY tables that are to be restored
     }
+
+    class StreamBatchWriter {
+
+        private final int maxBatchSize;
+        private final int maxWriteSize;
+        private final UUID streamId;
+
+        @Getter
+        private int totalNumSMREntries = 0;
+        @Getter
+        private int totalWriteSize = 0;
+
+        private final List<SMREntry> buffer = new ArrayList<>();
+        private int bufferWriteSize = 0;
+
+        StreamBatchWriter(int maxBatchSize, int maxWriteSize, UUID streamId) {
+            this.maxBatchSize = maxBatchSize;
+            this.maxWriteSize = maxWriteSize;
+            this.streamId = streamId;
+        }
+
+        public void batchWrite(List<SMREntry> smrEntries) {
+            if (smrEntries.isEmpty()) {
+                return;
+            }
+
+            Preconditions.checkState(buffer.size() < maxBatchSize);
+
+            for (SMREntry smrEntry : smrEntries) {
+                if (buffer.size() == maxBatchSize || bufferWriteSize + smrEntry.getSerializedSize() > maxWriteSize) {
+                    flushBuffer();
+                }
+                buffer.add(smrEntry);
+                bufferWriteSize += smrEntry.getSerializedSize();
+            }
+
+            // Keep buffer not full
+            if (buffer.size() == maxBatchSize || bufferWriteSize > maxWriteSize) {
```

Comment:
Given that you check if:
bufferWriteSize + smrEntry.getSerializedSize() > maxWriteSize
if bufferWriteSize > maxWriteSize, means a single smrEntry exceeded maxWriteSize, and the StreamsView.append would throw a WriteSizeException, is this right? 

Even if caller retries this will be in a loop forever.

Reply:
Thanks for reviewing.

Ideally if all writers correctly enforce the write size limit, there should be no such big SMREntry. Today it is the case because the max-write-size is set to 25MB for all JVMs. If somehow there is such big entry, compactor will also fail because it could not write that CP entry either. My point is if such corner case could happen the setup is in unrecoverable state anyways.